lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,523 @@
+    handleMuxError :: MuxError -> IO ()
+    handleMuxError = throwIO
+
+-- | A dummy network magic for a local cluster. When connecting to mainnet or
+-- testnet, this should match the underlying network's configuration.
+dummyNodeToClientVersion
+    :: (NodeToClientVersionData, CodecCBORTerm Text NodeToClientVersionData)
+dummyNodeToClientVersion =
+    ( NodeToClientVersionData { networkMagic = 0 }
+    , nodeToClientCodecCBORTerm
+    )
+
+-- | Construct a new network client handling 'ChainSync' and 'LocalTxSubmission'
+-- mini-protocols.
+newNetworkClient
+    :: forall m. (MonadThrow m, MonadST m, MonadTimer m)
+    => Tracer m String
+    -> ChainParameters
+    -> m (NetworkClient m, NetworkLayer m)
+newNetworkClient t params = do
+    chainSyncQueue <- atomically newTQueue
+
+    let client = OuroborosInitiatorApplication $ \pid -> \case
+            ChainSyncWithBlocksPtcl ->
+                chainSyncWithBlocks @m pid t params chainSyncQueue
+            LocalTxSubmissionPtcl ->
+                localTxSubmission @m pid t
+
+    let interface = NetworkLayer
+            { nextBlocks = \p -> ExceptT $ do
+                tvar <- newEmptyTMVarM
+                let request = ReqNextBlocks p (atomically . putTMVar tvar)
+                atomically $ writeTQueue chainSyncQueue request
+                atomically $ takeTMVar tvar
+            , networkTip = do
+                tvar <- newEmptyTMVarM
+                let request = ReqNetworkTip (atomically . putTMVar tvar)
+                atomically $ writeTQueue chainSyncQueue request
+                atomically $ takeTMVar tvar
+            }
+
+    return (client, interface)
+
+-- | Client for the 'Chain Sync' mini-protocol.
+--
+-- A corresponding 'Channel' can be obtained using a `MuxInitiatorApplication`
+-- constructor.
+chainSyncWithBlocks
+    :: forall m protocol peerId. ()
+    => (MonadThrow m, MonadST m, MonadSTM m, Show peerId)
+    => (protocol ~ ChainSync Block Point)
+    => peerId
+        -- ^ An abstract peer identifier for 'runPeer'
+    -> Tracer m String
+        -- ^ Base tracer for the mini-protocols
+    -> ChainParameters
+        -- ^ Some chain parameters necessary to encode/decode Byron 'Block'
+    -> TQueue m (ChainSyncRequest m)
+        -- ^ We use a 'TQueue' as a communication channel between the stateless
+        -- interface and the stateful connection. Requests are pushed to the queue
+        -- which handles them one-by-one.
+    -> Channel m ByteString
+        -- ^ A 'Channel' is a abstract communication instrument which
+        -- transports serialized messages between peers (e.g. a unix
+        -- socket).
+    -> m Void
+chainSyncWithBlocks pid t params queue channel =
+    runPeer trace codec pid channel (chainSyncClientPeer client)
+  where
+    trace :: Tracer m (TraceSendRecv protocol peerId DeserialiseFailure)
+    trace = contramap show t
+
+    codec :: Codec protocol DeserialiseFailure m ByteString
+    codec = codecChainSync
```

Comment:
it is quite low level, don't they have this parts already defined?

Reply:
hmmm... That's actually the high-level part of what they've available :sweat_smile: